jlm/hls/backend/rhls2firrtl/base-hls.hpp

```diff
@@ -72,63 +74,75 @@ class BaseHLS
   static std::string
   get_base_file_name(const llvm::RvsdgModule & rm);
 
-  std::vector<jlm::rvsdg::RegionArgument *>
-  get_mem_resps(const llvm::lambda::node * lambda)
+  /**
+   * Extracts all region arguments of the given kernel that represent memory responses.
+   * They can provide multiple values within a single execution of the region.
+   * @param lambda the lambda node holding the hls kernel
+   * @return the arguments that represent memory responses
+   */
+  std::vector<rvsdg::RegionArgument *>
+  get_mem_resps(const llvm::lambda::node & lambda)
   {
-    std::vector<jlm::rvsdg::RegionArgument *> mem_resps;
-    for (size_t i = 0; i < lambda->subregion()->narguments(); ++i)
+    std::vector<rvsdg::RegionArgument *> mem_resps;
+    for (auto arg : lambda.subregion()->Arguments())
     {
-      auto arg = lambda->subregion()->argument(i);
-      if (dynamic_cast<const jlm::hls::bundletype *>(&arg->type()))
-      {
-        mem_resps.push_back(lambda->subregion()->argument(i));
-      }
+      if (rvsdg::is<bundletype>(arg->type()))
+        mem_resps.push_back(arg);
     }
     return mem_resps;
   }
 
+  /**
+   * Extracts all region results of the given kernel that represent memory requests.
+   * They can take multiple values within a single execution of the region.
+   * @param lambda the lambda node holding the hls kernel
+   * @return the results that represent memory requests
+   */
   std::vector<rvsdg::RegionResult *>
-  get_mem_reqs(const llvm::lambda::node * lambda)
+  get_mem_reqs(const llvm::lambda::node & lambda)
   {
     std::vector<rvsdg::RegionResult *> mem_resps;
-    for (size_t i = 0; i < lambda->subregion()->nresults(); ++i)
+    for (auto result : lambda.subregion()->Results())
     {
-      if (dynamic_cast<const jlm::hls::bundletype *>(&lambda->subregion()->result(i)->type()))
-      {
-        mem_resps.push_back(lambda->subregion()->result(i));
-      }
+      if (rvsdg::is<bundletype>(result->type()))
+        mem_resps.push_back(result);
     }
     return mem_resps;
   }
 
-  std::vector<jlm::rvsdg::RegionArgument *>
-  get_reg_args(const llvm::lambda::node * lambda)
+  /**
+   * Extracts all region arguments of the given kernel that represent kernel inputs,
+   * which may include kernel arguments, state types, and context variables (always in that order).
+   * It will not return any arguments that represent memory responses.
+   * @param lambda the lambda node holding the hls kernel
+   * @return the arguments of the lambda that represent kernel inputs
+   */
+  std::vector<rvsdg::RegionArgument *>
+  get_reg_args(const llvm::lambda::node & lambda)
```

Comment:
These are harder to name, as it's basically get everything except for memory responses.
get_reg_args() sounds like one can get the arguments for a region, but this is only true for lambda.
Potential alternatives:

GetKernelArguments()  
# We should decide if we want to refer to the HLS function as the kernel, and in that case use it consistently.

GetHLSArguments()

Reply:
I kind of want to split this into separate functions for regular arguments, state arguments and context variables, to avoid code that implicitly relies on the ordering being consistent. It is probably a safe assumption, but it makes code harder to read. I was thinking a second PR could fix that